### PR TITLE
Implicits intro

### PR DESCRIPTION
**Pull request by hibikir**
First post for an intro to implicits, using typetags: the one example where the implicits are not added by any code.

### CODE REVIEW

**Comment by ryan-richt** [comment]
freaking awesome. Ship it! Srsly let's talk to Steph

**Comment by ddgenome** [comment]
The layout seems a bit weird on this post but I'm not sure why. Why is there a line break after "runtime." on the first line?

screen shot 2015-05-14 at 1 03 59 pm

File: _posts/2015-05-14-implicits-intro.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+So for this exercise, we'll see 
+<span class="get-inner-type">a little method</span>
+ that takes a List, 
+and returns the <span class="type-name">type name</span> 
+of the <span class="T"> contents of the list</span> .
+
+<div class="highlight"><pre>
+import scala.reflect.runtime.universe._
+
+def <span class="get-inner-type">getInnerType[<span class="T">T</span>]</span>(list:List[<span class="T">T</span>])(implicit tag:TypeTag[<span class="T">T</span>]) = tag.<span class="type-name">tpe.toString</span>
```
Comment (ddgenome) [inline _posts/2015-05-14-implicits-intro.md]: Should that last bit be `type.toString`?

File: _posts/2015-05-14-implicits-intro.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+an implicit TypeTag or ClassPath parameter for all real classes (as opposed to generics, which we'll cover in a minute).
+
+The context of the call to getInnerType knows that list is a List[String], so the compiler fills in the implicit for us.
+The compiler fills the implicit as if we had called the method like this:
+
+<div class="highlight"><pre>
+  import scala.reflect.runtime.universe._
+
+  val stringList: List[String] = List("A")   
+  val stringName = <span class="get-inner-type">getInnerType</span>(stringList)(typeTag(String))
```
Comment (ddgenome) [inline _posts/2015-05-14-implicits-intro.md]: Is it possible to highlight the new bit, `(typeTag(String))`?

File: _posts/2015-05-14-implicits-intro.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+{% endhighlight %}
+
+Great, we defeated erasure! But how did this work? How did that implicit TypeTag get there? Compiler magic!
+
+The easiest way to think about implicit parameters is that they are extra parameters to a function that can be populated
+by the compiler instead of being passed manually. In the case of TypeTags and ClassTags,
+we do not have to do anything to make them work: the compiler will always be able to provide
+an implicit TypeTag or ClassPath parameter for all real classes (as opposed to generics, which we'll cover in a minute).
+
+The context of the call to getInnerType knows that list is a List[String], so the compiler fills in the implicit for us.
```
Comment (ddgenome) [inline _posts/2015-05-14-implicits-intro.md]: Should you backtick `getInnerType` and `List[String]`?

Also, you should probably change "list" to "`stringList`".

File: _posts/2015-05-14-implicits-intro.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+Therefore, it cannot provide the implicit parameter. Only the callers to gratuitousIntermediateMethod will know
+what the type is, and therefore be able to provide the TypeTag.
+
+So to make this work, the intermediate method needs to request the TypeTag itself, 
+so that the compiler can pass the TypeTag down to getInnerType:
+
+<div class="highlight"><pre>
+  import scala.reflect.runtime.universe._
+
+  def <span class="grm">gratuitousIntermediateMethod</span>[<span class="T">T</span>](list:List[<span class="T">T</span>])(implicit tag :TypeTag[<span class="T">T</span>]) = getInnerType(list)
```
Comment (ddgenome) [inline _posts/2015-05-14-implicits-intro.md]: Break this so it does not get broken by the formatting?
screen shot 2015-05-14 at 1 14 05 pm

**Comment by ddgenome** [comment]
Should this be using `{% highlight scala %}` throughout?

File: _posts/2015-05-14-implicits-intro.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+</pre></div>
+
+This works just like our example without the gratuitous method.
+
+<div class="highlight"><pre>
+a list of java.lang.String
+</pre></div>
+
+We need parameters that carry the TypeTag from where the type is concrete to where the typeTag is needed.
+This kind of pattern, carrying implicits over, happens with other kinds of implicit parameter:
```
Comment (ddgenome) [inline _posts/2015-05-14-implicits-intro.md]: "implicit parameters:"?

**Comment by ddgenome** [comment]
While the material is tough, I think the presentation is really clear. We should get this post out soon!

**Comment by ddgenome** [comment]
I made a commit with all my suggestions. I hope I didn't mess anything up too badly.

**Pull request merged**